src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControlCheck.java

```diff
@@ -79,15 +79,17 @@
     private boolean processCurrentFile;
 
     /** The root package controller. */
-    private ImportControl root;
+    private AbstractImportControl root;
```

Comment:
it is not clear why type is changed.
please explain how root could be FileImportControl ?
root is `<import-control` some synonym for `<subpackage`

Loader on `getRoot()` should return exact type - PkgImportControl .

Reply:
Type changed because there is no `ImportControl` anymore. It was renamed to `AbstractImportControl` and pieces split apart into `PkgImportControl`.

`ImportControlLoader` returns what is on the top of it's stack, which is a list of `AbstractImportControl` because it needs to handle both node types.

We can cast the return to `PkgImportControl` but why don't we leave it as abstract incase we add new root nodes?

Reply:
it is PkgImportControl.

> ImportControlLoader returns what is on the top of it's stack

can it be different form PkgImportControl ? can you give me example of meaningful config ?
As term is root - it is possible to have only one root element.
More over:
```
$ head import-control.xml 
<?xml version="1.0"?>
<!DOCTYPE import-control PUBLIC
    "-//Puppy Crawl//DTD Import Control 1.3//EN"
    "http://checkstyle.sourceforge.net/dtds/import_control_1_3.dtd">

<import-control pkg="com.my">
  <disallow class="java.io.*"/>
</import-control>

<import-control pkg="com.puppycrawl.tools.checkstyle">

$ xmllint import-control.xml 
import-control.xml:10: parser error : Extra content at the end of the document
<import-control pkg="com.puppycrawl.tools.checkstyle">
```

> We can cast the return to PkgImportControl but why don't we leave it as abstract incase we add new root nodes?

Such classes are private implementation, so following a rule "make it extendable in future" will not work well, not you neighter I know future. It is better to be precise now.
Two import-control nodes in one file is separate discussion/issue.

Reply:
> Two import-control nodes in one file is separate discussion/issue.
> As term is root - it is possible to have only one root element.

I was never advocating 2 root elements. root right now is `PkgImportControl`. I was saying in the future we could have `XyzImportControl` and it could be the root over `PkgImportControl` through user config.
If you look at `ImportControlLoader`, it gets root from the field named `stack`. That is what I was referring to.

Reply:
> If you look at ImportControlLoader, it gets root from the field named stack. That is what I was referring to.

I saw, but lets address stack usage in separate issue if required. 
Right now , out of Loader should come one element and it is PkgImportControl. 
When we come to XyzImportControl we will refactor appropriately, future will come with new requirements that we can not predict now. In future on the moment of refactoring it is better to clearly see what code do/support now, rather than guess what for such option exists and for whom it was done and .... .

Reply:
Done.